### PR TITLE
Support loading weights from cache for GPT-OSS 

### PR DESCRIPTION
**Pull request by sankarmanoj-tt**
### Problem description
Loading the HF Model Weights takes a long time due to type conversion from MXFP4 to BFloat8. This loaded is redundant if the cache is filled. 

### What's changed
Enable support for only loading the weights from cache

### Checklist

- [ ] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch:smanoj/gptoss_optional_state_dict_load)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:smanoj/gptoss_optional_state_dict_load)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:smanoj/gptoss_optional_state_dict_load)
- [ ] New/Existing tests provide coverage for changes
- [ ] (Optional) Ran [clang-tidy code analysis](https://github.com/tenstorrent/tt-metal/actions/workflows/code-analysis.yaml) on the PR branch to catch linting errors early

#### Model tests

If your changes cover model-related code, you should run tests corresponding to affected models and platforms (Single card, T3K, Galaxy). "Choose your pipeline" workflows facilitate running multiple kinds of tests in a single run. Each offers `models-mandatory` and `models-extended` presets.
The former includes a minimal set of tests, to be run always. The latter extends that with additional ones - use your best judgement in deciding which is the most appropriate for your PR.

- [ ] [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:smanoj/gptoss_optional_state_dict_load)
  - [ ] `models-mandatory` preset (runs: [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) and [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:smanoj/gptoss_optional_state_dict_load)
  - [ ] `models-mandatory` preset (runs: [Unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:smanoj/gptoss_optional_state_dict_load)
  - [ ] `models-mandatory` preset (runs: [Quick tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

### CODE REVIEW

**Comment by sraizada-tt** [comment]
## Critical Issues - Must Fix Before Merge

### 🔴 Issue 1: `o_proj_cache_suffix` logic broken

**File:** `tt/attention/weights.py` line ~247

**Problem:** Removed the `o_proj_pad_size > 0` check from cache suffix logic:

```python
# Current (BROKEN):
o_proj_cache_suffix = f"_padded" if mesh_config.tp > 1 else ""

# Should be:
o_proj_cache_suffix = f"_padded" if o_proj_pad_size > 0 and mesh_config.tp > 1 else ""
```

Now adds `"_padded"` suffix even when padding wasn't applied (`o_proj_pad_size == 0`), causing cache key mismatch.

---

### 🔴 Issue 2: `model_path` can be None, then used in string checks

**File:** `demo/text_demo.py` lines 478-495

**Problem:**
```python
model_path = os.getenv("HF_MODEL", None)  # Can be None!
...
if "120b" in model_path and mesh_device.shape[0] == 1:  # TypeError if model_path is None
```

**Fix:** Add None guards to all 3 model_path checks:
```python
if model_path and "120b" in model_path and mesh_device.shape[0] == 1:
    ...
if model_path and "120b" in model_path and mesh_device.shape[0] == 4:
    ...
if model_path and "20b" in model_path and mesh_device.shape[0] == 4:
    ...
```

---

### 🔴 Issue 3: `state_dict` check inconsistency

**Files:** `demo/text_demo.py` + `tt/common.py`

**Problem:** Demo passes empty dict `{}` but `common.py` checks `is None`:

```python
# Demo (line 432):
state_dict = {}  # Empty dict when load_model=False

# common.py (line 51):
if state_dict is None:  # Only catches None, not {}
    state_dict = gpt_oss_model_args.load_state_dict(...)
```

**Result:** Will still load state_dict even when `load_model=False` because `{} is None` → False.

**Fix option A (simpler):** Revert `common.py` to falsy check:
```python
if not state_dict:  # Catches both None and {}
```

**Fix option B:** Change demo to pass `None` instead of `{}`.

---

## Additional Notes

- Cache key format changed (removed shape suffix from expert biases) — this is a **breaking change** that invalidates existing caches. Should document in PR description.
- `load_model = False` is hardcoded in demo. Consider making it controllable via env var or pytest parameter.

---

Please fix these 3 critical issues and **run CI** to verify the changes work correctly. Thanks!

File: models/demos/gpt_oss/demo/text_demo.py [hunk 1]
```diff
@@ -420,8 +421,30 @@ def test_gpt_oss_demo(
     run_in_ci,
     is_ci_env,
     request,
-    state_dict,
 ):
+    # Set to False to optimize for execution time.
+    # If True, will always try to load HF weights if path is set.
+    load_model = False
+    if load_model:
+        model_path = os.getenv("HF_MODEL", None)
```
[the remark this answers is not in the document]
Comment (sraizada-tt) [reply models/demos/gpt_oss/demo/text_demo.py]: @sankarmanoj-tt can you address this

**Comment by atupe-tt** [comment]
Galaxy CI passing - https://github.com/tenstorrent/tt-metal/actions/runs/23095439623

**Review by sraizada-tt — APPROVED** [review]

**Review by uaydonat — APPROVED** [review]
